src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorHtmlPublishDiagnosticsInterceptor.cs

```diff
@@ -0,0 +1,195 @@
+            }
+
+            var razorDocumentUri = RazorLSPConventions.GetRazorDocumentUri(diagnosticParams.Uri);
+
+            if (!_documentManager.TryGetDocument(razorDocumentUri, out var razorDocumentSnapshot))
+            {
+                return CreateDefaultResponse(token);
+            }
+
+            if (!razorDocumentSnapshot.TryGetVirtualDocument<HtmlVirtualDocumentSnapshot>(out var htmlDocumentSnapshot))
+            {
+                return CreateDefaultResponse(token);
+            }
+
+            // Ensure we're working with the virtual document specified in the diagnostic params
+            if (!htmlDocumentSnapshot.Uri.Equals(diagnosticParams.Uri))
+            {
+                return CreateDefaultResponse(token);
+            }
+
+            // Note; this is an `interceptor` & not a handler, hence 
+            // it's possible another interceptor mutates this request 
+            // later in the toolchain. Such an interceptor would likely 
+            // expect a `__virtual.html` suffix instead of `.razor`.
+            diagnosticParams.Uri = razorDocumentUri;
+
+            // Return early if there aren't any diagnostics to process
+            if (diagnosticParams.Diagnostics?.Any() != true)
+            {
+                return CreateResponse(diagnosticParams);
+            }
+
+            var synchronized = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync(
+                razorDocumentSnapshot.Version,
+                htmlDocumentSnapshot,
+                cancellationToken).ConfigureAwait(false);
+            if (!synchronized)
+            {
+                // Could not synchronize, we'll clear out the diagnostics as we don't
+                // know whether or not they're still valid in the current state of the
+                // document.
+                return CreateResponse(diagnosticParams, clearDiagnostics: true);
+            }
+
+            var unmappedDiagnostics = diagnosticParams.Diagnostics;
+            var filteredDiagnostics = unmappedDiagnostics.Where(d => !CanDiagnosticBeFiltered(d));
+            if (!filteredDiagnostics.Any())
+            {
+                return CreateResponse(diagnosticParams, clearDiagnostics: true);
+            }
+
+            var rangesToMap = filteredDiagnostics.Select(r => r.Range).ToArray();
+            var mappingResult = await _documentMappingProvider.MapToDocumentRangesAsync(
+                RazorLanguageKind.Html,
+                razorDocumentUri,
+                rangesToMap,
+                LanguageServerMappingBehavior.Inclusive,
+                cancellationToken).ConfigureAwait(false);
+
+            if (mappingResult == null || mappingResult.HostDocumentVersion != razorDocumentSnapshot.Version)
+            {
+                return CreateResponse(diagnosticParams, clearDiagnostics: true);
+            }
+
+            var mappedDiagnostics = new List<Diagnostic>(filteredDiagnostics.Count());
```

Comment:
I wouldn't specify the capacity here since you technically aren't promising that you're going to fill out the list since you'll be only accepting the ranges that are mappable I presume?

Reply:
Yes, but the assumption was that you'll likely be filling out most/all of the list every-time. Especially with HTML diagnostics, I don't believe it should be possible for the mapping to fail.